### PR TITLE
etcdserver: clean up start and stop logic of raft

### PR DESCRIPTION
**Pull request by yichengq**
kill TODO and make it more readable.

follows #3077

### CODE REVIEW

File: etcdserver/raft.go [hunk 1]
```diff
@@ -176,6 +184,11 @@ func (r *raftNode) apply() chan apply {
 }
 
 func (r *raftNode) stop() {
+	r.stopped <- struct{}{}
+	<-r.done
+}
+
+func (r *raftNode) close() {
```
Comment (xiang90) [inline etcdserver/raft.go]: can we rename this to shutdown or something else with a comment? this confuses me with `stop`.

File: etcdserver/raft.go [hunk 2]
```diff
@@ -108,10 +108,18 @@ type raftNode struct {
 	done    chan struct{}
 }
 
+func (r *raftNode) start(s *EtcdServer) {
```
Comment (xiang90) [inline etcdserver/raft.go]: can we add a comment about start? also is it possible to combine start and run? Ideally we should get rid of the passed in server structure.

**Comment by yichengq** [comment]
Addressed the comments.

> Ideally we should get rid of the passed in server structure.

I tried, then it needs some line `srv.r.s = srv`, which is not easy to read then. It could be wrapped in some member function to be solved though.

**Comment by xiang90** [comment]
> I tried, then it needs some line srv.r.s = srv, which is not easy to read then. It could be wrapped in some member function to be solved though.

I mean `Ideally we should get rid of the passed in server structure`. I do not quite understand what do you mean by srv.r.s = srv. If we get rid of server structure, then there is no s in r.

**Comment by yichengq** [comment]
@xiang90 Agree with your point. I have added a TODO for getting rid of server structure.

File: etcdserver/raft.go [hunk 3]
```diff
@@ -176,6 +186,11 @@ func (r *raftNode) apply() chan apply {
 }
 
 func (r *raftNode) stop() {
+	r.stopped <- struct{}{}
+	<-r.done
+}
+
+func (r *raftNode) shutdown() {
```
Comment (xiang90) [inline etcdserver/raft.go]: onStop?

**Comment by xiang90** [comment]
LGTM.

File: etcdserver/raft.go (hunk 3 above)
Comment (yichengq) [inline etcdserver/raft.go]: changed.

**Pull request merged**